File: packages/pz_agent_core/src/pz_agent_core/actions/adapters/medical.py
"""``medical.bandage``: dressing one wound, and proving it.

The postcondition is read off the body part the command named: the wound the
character was losing blood from is no longer reported as bleeding. Nothing here
writes a flag on the damage model, and nothing here treats "the dressing left
the inventory" as proof — a bandage that was used up and a bandage that was
dropped look identical from the item list, and only one of them stopped the
bleeding.

Two things this adapter deliberately does not do.

**It does not choose.** Which part and which dressing are
:func:`pz_agent_core.policy.medical.select_treatment`'s decision, made
deterministically and separately tested, and both arrive in the command.
Duplicating the triage here would create a second copy of the rule that refuses
a dirty dressing while a clean one is in reach — and a second copy is a copy no
test compares against the first.

**It does not treat what it cannot observe.** The observation reports a wound's
``bleeding`` flag; it reports no bandage flag at all. Dressing a wound that is
not bleeding is therefore refused rather than attempted, because the only thing
that could follow is an ack this side would have to take on trust. That is a
narrower action than the mod's — ``adapters/Medical.lua`` is willing to bandage
anything the game calls a wound — and the narrowing is on purpose: this side
only sends what it can check.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import ClassVar, Final

from ...capabilities.probes import MEDICAL_BANDAGE
from ...policy.medical import Dressing, wound_body_part
from ...protocol import (
    ActionName,
    Command,
    JsonDict,
    Observation,
    PlayerState,
    ReasonCode,
    RefKind,
    RiskClass,
    Wound,
)
from ..adapter import Evidence, PreconditionFailed
from .common import (
    check_args,
    find_by_identity,
    identity_of,
    read_ref,
    require_inventory,
    resolve_item,
)
from .consume import require_in_main

__all__ = [
    "BODY_PARTS",
    "DEFAULT_BANDAGE_POLL_MS",
    "DEFAULT_BANDAGE_TIMEOUT_MS",
    "BandageAdapter",
]

#: Body parts a command may name, spelled the way ``BodyPartType`` spells them
#: and the way the mod's wound references carry them.
#:
#: Closed for two reasons. A build that grows a part treats it only once it is
#: listed here, which is the direction to be wrong in; and ``body_part`` is used
#: to look a wound up by name, so an unchecked value would be a lookup on a
#: string that arrived from a model.
BODY_PARTS: Final[frozenset[str]] = frozenset(
    {
        "Head",
        "Neck",
        "Torso_Upper",
        "Torso_Lower",
        "Groin",
        "UpperArm_L",
        "UpperArm_R",
        "ForeArm_L",
        "ForeArm_R",
        "Hand_L",
        "Hand_R",
        "UpperLeg_L",
        "UpperLeg_R",
        "LowerLeg_L",
        "LowerLeg_R",
        "Foot_L",
        "Foot_R",
    }
)

DEFAULT_BANDAGE_TIMEOUT_MS: Final = 30_000
DEFAULT_BANDAGE_POLL_MS: Final = 250


def _wound_on(player: PlayerState, body_part: str) -> Wound | None:
    """The wound reported for one body part, or None when none is.

    A part with no wound in the list is not injured — which after a successful
    dressing is exactly what the observation should say, and is why the absence
    is treated as an outcome here rather than as missing data.
    """
    return next((w for w in player.wounds if wound_body_part(w) == body_part), None)


@dataclass(frozen=True, slots=True)
class _BandageSpec:
    """The part to dress and the dressing to dress it with.

    Both are required, unlike on the mod side where either may be omitted. The
    reason is the one in the module docstring: with the choice made by the
    policy and carried in the command, ``verify`` knows which part to look at
    even after the wound it was chosen for has stopped bleeding — re-deriving
    "the worst bleeding part" from the world as it is *now* would go looking at
    a different part precisely when the action worked.
    """

    body_part: str
    item_ref: str

    @classmethod
    def parse(cls, command: Command) -> _BandageSpec:
        check_args(command, allowed=("body_part", "item_ref"), required=("body_part", "item_ref"))
        raw = command.args.get("body_part")
        if not isinstance(raw, str) or raw not in BODY_PARTS:
            raise PreconditionFailed(
                f"body_part must be one of the {len(BODY_PARTS)} parts this character has",
                reason_code=ReasonCode.INVALID_ARGUMENT,
                evidence={"body_part": raw if isinstance(raw, str) else None},
            )
        return cls(body_part=raw, item_ref=read_ref(command, "item_ref", kind=RefKind.ITEM))

    def as_args(self) -> JsonDict:
        return {"body_part": self.body_part, "item_ref": self.item_ref}


@dataclass(frozen=True, slots=True)
class BandageAdapter:
    """``medical.bandage``: dress one bleeding wound with one carried dressing."""

    timeout_ms: int = DEFAULT_BANDAGE_TIMEOUT_MS
    poll_interval_ms: int = DEFAULT_BANDAGE_POLL_MS

    name: ClassVar[ActionName] = ActionName.MEDICAL_BANDAGE
    risk: ClassVar[RiskClass] = RiskClass.P2
    required_capability: ClassVar[str | None] = MEDICAL_BANDAGE

    def validate(self, command: Command, observation: Observation) -> None:
        spec = _BandageSpec.parse(command)
        inventory = require_inventory(observation)
        item = resolve_item(inventory, spec.item_ref)
        if Dressing.classify(item) is None:
            raise PreconditionFailed(
                f"{item.display_name} is not a dressing",
                reason_code=ReasonCode.INVALID_ARGUMENT,
                evidence={"item_ref": item.ref, "full_type": item.full_type},
            )
        # Deliberately not the triage filters: whether a dirty dressing may be
        # used, and whether a reserved one may be spent, are policy.medical's
        # decisions and are already made by the time a reference reaches here.
        wound = _wound_on(observation.player, spec.body_part)
        if wound is None:
            raise PreconditionFailed(
                f"no wound is reported on {spec.body_part}",
                reason_code=ReasonCode.PRECONDITION_FAILED,
                evidence={"body_part": spec.body_part},
            )
        if not wound.bleeding:
            raise PreconditionFailed(
                f"{spec.body_part} is not bleeding, and the observation reports no "
                "dressing state, so applying one could not be verified",
                reason_code=ReasonCode.PRECONDITION_FAILED,
                evidence={
                    "body_part": spec.body_part,
                    "wound_ref": wound.ref,
                    "wound_kind": wound.kind,
                },
            )
        require_in_main(inventory, item)

    def build_args(self, command: Command, observation: Observation) -> JsonDict:
        return _BandageSpec.parse(command).as_args()

    def verify(self, command: Command, before: Observation, after: Observation) -> Evidence | None:
        spec = _BandageSpec.parse(command)
        was = _wound_on(before.player, spec.body_part)
        if was is None or not was.bleeding:
            # Nothing was bleeding there to begin with, so nothing about the
            # part afterwards can prove this command did anything.
            return None
        now = _wound_on(after.player, spec.body_part)
        if now is not None and now.bleeding:
            return None
        identity = identity_of(spec.item_ref)
        used = after.inventory is not None and not find_by_identity(after.inventory, identity)
        return Evidence(
            kind="bleeding_stopped",
            observation_seq=after.seq,
            observed={
                "body_part": spec.body_part,
                "wound_ref": was.ref,
                "wound_kind": was.kind,
                "bleeding_before": True,
                "bleeding_after": False if now is None else now.bleeding,
                "severity_before": was.severity,
                "severity_after": None if now is None else now.severity,
                "wound_still_reported": now is not None,
                "item_ref": spec.item_ref,
                # Context, never proof: a dressing can leave the inventory by
                # being dropped, so this says what happened to it and claims
                # nothing about why the bleeding stopped.
                "dressing_consumed": used,
            },
        )

File: pz-mod/42/media/lua/client/PZAgent/adapters/Medical.lua
--[[
PZAgent.Adapters.Medical -- medical.bandage.

Invariant: the dressing is applied by the game's own ISApplyBandage against a
real BodyPart object, and the postcondition is read back off that body part --
bleeding false, or bandaged true. No flag on the body damage model is ever
written here. Writing `bleeding = false` would make every ack succeed and every
character die of a wound the agent had recorded as treated.

Which part and which dressing are both decided deterministically, and both are
decided from the observation rather than from the command: the worst bleeding
part wins, ties break on the part's own name, and dressings come from a closed
ordered list with the dirty ones last. A caller may name either explicitly, and
naming a part that is not bleeding is refused rather than quietly redirected.

The constructor's argument order is vanilla's documented
`(doctor, patient, item, bodyPart)`. There is no Build 42.20 here to check that
against; a build that orders them differently produces a bandage that does not
land, and the postcondition below is what refuses to call that a success.
]]

-- Load-order guard, live-proven 2026-08-08 on Build 42.20.2: the engine walked
-- adapters/ in an order that ran this file before Toolkit.lua. The statement
-- form is deliberate -- the paren form is banned as dynamic loading -- and the
-- test harness pre-resolves this module, so there the require is a no-op.
require "PZAgent/adapters/Toolkit"

PZAgent = PZAgent or {}
PZAgent.Adapters = PZAgent.Adapters or {}

local Medical = {}
PZAgent.Adapters.Medical = Medical

local function toolkit()
  return PZAgent.Adapters.Toolkit
end

--- Dressings this adapter will pick on its own, best first.
---
--- Closed and ordered on purpose. Dirty dressings are last because they carry
--- an infection risk the agent should only take when there is nothing else;
--- anything outside this list has to be named explicitly with `item_ref`, so a
--- mod-added dressing is a decision the caller makes rather than one made here
--- by a name match.
Medical.DRESSINGS = {
  "Base.Bandage",
  "Base.RippedSheets",
  "Base.DenimStrips",
  "Base.LeatherStrips",
  "Base.RippedSheetsDirty",
  "Base.DenimStripsDirty",
}

--- Body parts a caller may name, spelled the way BodyPartType spells them.
---
--- Closed for the same reason the dressings are, and for one more: `body_part`
--- is used as a key into the observed body, so a name that arrived from the
--- wire unchecked would be a lookup on data an attacker chose. A build that
--- grows a part treats it only once it is listed here, which is the direction
--- to be wrong in.
Medical.BODY_PARTS = {
  Head = true,
  Neck = true,
  Torso_Upper = true,
  Torso_Lower = true,
  Groin = true,
  UpperArm_L = true,
  UpperArm_R = true,
  ForeArm_L = true,
  ForeArm_R = true,
  Hand_L = true,
  Hand_R = true,
  UpperLeg_L = true,
  UpperLeg_R = true,
  LowerLeg_L = true,
  LowerLeg_R = true,
  Foot_L = true,
  Foot_R = true,
}

Medical.TIMEOUT_MS = 30000
Medical.POLL_MS = 250

local REQUIRES = { "ISApplyBandage", "ISApplyBandage.new", "ISTimedActionQueue.add" }

local BANDAGE_ARGS = { "body_part", "item_ref" }

--- Argument kinds, spelled out rather than read from
--- PZAgent.CommandDispatcher.ARG, because the dispatcher is a sibling file and
--- nothing here may dereference a module at load time. A typo costs a refused
--- registration at load rather than a loosened check at dispatch.
local ARG = {
  REF = "ref",
  ENUM = "enum",
}

--- The worst bleeding body part in a snapshot, or nil.
---
--- Severity first, then the part's own name, so two equally hurt arms are always
--- treated in the same order and a retry treats the same one.
function Medical.worstBleeding(snapshot)
  if type(snapshot) ~= "table" or type(snapshot.body) ~= "table" then
    return nil
  end
  local best = nil
  local bestName = nil
  for name, part in pairs(snapshot.body) do
    if part.bleeding == true then
      local severity = part.severity or 0
      if
        best == nil
        or severity > (best.severity or 0)
        or (severity == (best.severity or 0) and name < bestName)
      then
        best = part
        bestName = name
      end
    end
  end
  if best == nil then
    return nil
  end
  return bestName, best
end

--- The live BodyPart object behind a snapshot entry.
---
--- Re-read rather than cached: the snapshot carries the index the part was at,
--- and the object itself must come from the engine at the moment it is used.
function Medical.bodyPartAt(player, index)
  local Toolkit = toolkit()
  local ok, body = Toolkit.call(player, "getBodyDamage")
  if not ok or body == nil then
    return nil, "IsoPlayer.getBodyDamage"
  end
  local okParts, parts = Toolkit.call(body, "getBodyParts")
  if not okParts or parts == nil then
    return nil, "BodyDamage.getBodyParts"
  end
  local part = Toolkit.listGet(parts, index)
  if part == nil then
    return nil, "BodyDamage.getBodyParts().get"
  end
  return part
end

--- The dressing to use: the one the caller named, or the best one carried.
local function chooseDressing(ctx, itemRef)
  local Toolkit = toolkit()
  local Inventory = PZAgent.Adapters.Inventory
  if itemRef ~= nil then
    local item, code, detail = Toolkit.resolveItem(ctx, itemRef)
    if item == nil then
      return nil, code, detail
    end
    return { ref = itemRef, item = item }
  end
  for index = 1, #Medical.DRESSINGS do
    local hits = Inventory.search(ctx, { full_type = Medical.DRESSINGS[index], limit = 1 })
    if #hits == 1 and hits[1].ref ~= nil then
      local item, code, detail = Toolkit.resolveItem(ctx, hits[1].ref)
      if item == nil then
        return nil, code, detail
      end
      return { ref = hits[1].ref, item = item, full_type = Medical.DRESSINGS[index] }
    end
  end
  return nil,
    Toolkit.reasons().PRECONDITION_FAILED,
    "the character carries none of the dressings this adapter will choose on its own"
end

Medical.chooseDressing = chooseDressing

local function bandageSpec(args, ctx)
  local Toolkit = toolkit()
  local reasons = Toolkit.reasons()
  local checked, checkCode, checkDetail = Toolkit.checkArgs(args, BANDAGE_ARGS, {})
  if checked == nil then
    return nil, checkCode, checkDetail
  end
  local snapshot = Toolkit.snapshot(ctx)
  -- Emptiness via PZAgent.Compat, not the global `next`, which the 2026-08-08
  -- live run proved Kahlua does not provide; hasEntries also answers false for
  -- a non-table, preserving the type check this line used to spell out.
  if not PZAgent.Compat.hasEntries(snapshot.body) then
    return Toolkit.unavailable("BodyDamage.getBodyParts")
  end
  local partName, part
  if args.body_part ~= nil then
    local named, namedCode, namedDetail = Toolkit.readText(args, "body_part", { maximum = 64 })
    if named == nil then
      return nil, namedCode, namedDetail
    end
    part = snapshot.body[named]
    if part == nil then
      return nil, reasons.INVALID_ARGUMENT, string.format("this character has no body part called %q", named)
    end
    if part.bleeding ~= true then
      return nil, reasons.PRECONDITION_FAILED, string.format("%s is not bleeding", named)
    end
    partName = named
  else
    partName, part = Medical.worstBleeding(snapshot)
    if partName == nil then
      return nil, reasons.PRECONDITION_FAILED, "nothing on this character is bleeding"
    end
  end
  local itemRef
  if args.item_ref ~= nil then
    local given, givenCode, givenDetail = Toolkit.readRef(args, "item_ref", PZAgent.Refs.KIND.ITEM, ctx)
    if given == nil then
      return nil, givenCode, givenDetail
    end
    itemRef = given
  end
  return { body_part = partName, part = part, item_ref = itemRef, bleeding_before = true }
end

Medical.bandageSpec = bandageSpec

local Bandage = nil

local function bandageValidate(_, args, ctx)
  local Toolkit = toolkit()
  local required, requiredCode, requiredDetail = Toolkit.requireSymbols(Bandage.requires)
  if required == nil then
    return nil, requiredCode, requiredDetail
  end
  local spec, code, detail = bandageSpec(args, ctx)
  if spec == nil then
    return nil, code, detail
  end
  local dressing, dressingCode, dressingDetail = chooseDressing(ctx, spec.item_ref)
  if dressing == nil then
    return nil, dressingCode, dressingDetail
  end
  spec.dressing = dressing
  Toolkit.state(ctx).bandage = spec
  return true
end

--- ISApplyBandage takes the dressing out of the character's own inventory, so a
--- roll of bandages at the bottom of a backpack is fetched first.
local function bandagePrepare(_, args, ctx)
  local Toolkit = toolkit()
  local code, detail = Toolkit.interruption(ctx)
  if code ~= nil then
    return nil, code, detail
  end
  local spec, specCode, specDetail = bandageSpec(args, ctx)
  if spec == nil then
    return nil, specCode, specDetail
  end
  local dressing, dressingCode, dressingDetail = chooseDressing(ctx, spec.item_ref)
  if dressing == nil then
    return nil, dressingCode, dressingDetail
  end
  local outcome, moveCode, moveDetail = PZAgent.Adapters.Inventory.bringToMain(ctx, dressing.ref)
  if outcome == nil then
    return nil, moveCode, moveDetail
  end
  Toolkit.state(ctx).fetched = outcome == "moving"
  return true
end

local function bandageStart(_, args, ctx)
  local Toolkit = toolkit()
  local code, detail = Toolkit.interruption(ctx)
  if code ~= nil then
    return nil, code, detail
  end
  local spec, specCode, specDetail = bandageSpec(args, ctx)
  if spec == nil then
    return nil, specCode, specDetail
  end
  local dressing, dressingCode, dressingDetail = chooseDressing(ctx, spec.item_ref)
  if dressing == nil then
    return nil, dressingCode, dressingDetail
  end
  local part, missing = Medical.bodyPartAt(ctx.player, spec.part.index)
  if part == nil then
    return Toolkit.unavailable(missing)
  end
  local action, actionCode, actionDetail =
    Toolkit.construct("ISApplyBandage", ctx.player, ctx.player, dressing.item, part)
  if action == nil then
    return nil, actionCode, actionDetail
  end
  local queued, queueCode, queueDetail = Toolkit.enqueue(ctx, action)
  if queued == nil then
    return nil, queueCode, queueDetail
  end
  Toolkit.state(ctx).applied_ref = dressing.ref
  return true
end

local function bandageVerify(_, before, after, args, ctx)
  local Toolkit = toolkit()
  local reasons = Toolkit.reasons()
  -- The spec is re-derived from the *before* snapshot rather than from the
  -- world as it is now: once the wound stops bleeding, "the worst bleeding
  -- part" is a different part, and verify would go looking at the wrong one.
  local partName = Toolkit.state(ctx).bandage ~= nil and Toolkit.state(ctx).bandage.body_part or args.body_part
  if partName == nil then
    local chosen = Medical.worstBleeding(before)
    partName = chosen
  end
  if partName == nil then
    return nil, reasons.POSTCONDITION_FAILED, "no bleeding body part was recorded for this command"
  end
  local was = type(before) == "table" and type(before.body) == "table" and before.body[partName] or nil
  local now = type(after) == "table" and type(after.body) == "table" and after.body[partName] or nil
  if now == nil then
    return Toolkit.unavailable("BodyDamage.getBodyParts")
  end
  local observed = {
    kind = "wound_dressed",
    body_part = partName,
    bleeding_before = was ~= nil and was.bleeding or nil,
    bleeding_after = now.bleeding,
    bandaged_after = now.bandaged,
    item_ref = Toolkit.state(ctx).applied_ref,
  }
  if now.bleeding == false or now.bandaged == true then
    return observed
  end
  return nil,
    reasons.POSTCONDITION_FAILED,
    string.format("%s is still bleeding and still reports no dressing", partName)
end

Bandage = toolkit().declare({
  name = "medical.bandage",
  capability = toolkit().CAPABILITY.MEDICAL_BANDAGE,
  requires = REQUIRES,
  timeout_ms = Medical.TIMEOUT_MS,
  poll_interval_ms = Medical.POLL_MS,
  args = {
    -- Neither is required: with nothing named the adapter treats the worst
    -- bleeding part with the best dressing carried, which is the deterministic
    -- choice the file exists to make.
    body_part = { type = ARG.ENUM, values = Medical.BODY_PARTS },
    item_ref = { type = ARG.REF, kinds = { item = true } },
  },
  validate = bandageValidate,
  prepare = bandagePrepare,
  begin = bandageStart,
  verify = bandageVerify,
})

Medical.Bandage = Bandage
toolkit().register(Bandage)

return Medical
